split.go

```diff
@@ -0,0 +1,113 @@
+// too long IRC messages. The function is passed an event which must be
+// split into multiple and the maximum event length.
+type splitFunc func(event *Event, maxLen int) []*Event
+
+var splitFuncs = map[string]splitFunc{
+	PRIVMSG: splitPRIVMSG,
+}
+
+// maxHostLen returns the maximum possible length of a server message
+// prefix as defined by the following ABNF in RFC 2812:
+//
+//   [ ":" prefix SPACE ]
+//
+func maxPrefixLen(client *Client) int {
+	var nicklen, hostlen int
+	var err error
+
+	nickval, success := client.GetServerOption("MAXNICKLEN")
+	if success {
+		nicklen, err = strconv.Atoi(nickval)
+	}
+	if !success || err != nil {
+		nicklen = 10 // common value according to irssi
+	}
+
+	hostval, success := client.GetServerOption("HOSTLEN")
+	if success {
+		hostlen, err = strconv.Atoi(hostval)
+	}
+	if !success || err != nil {
+		hostlen = 63 // common value according to irssi
+	}
+
+	// From RFC 2812:
+	//
+	//   prefix = servername / ( nickname [ [ "!" user ] "@" host ] )
+	//
+	// The code here assumes the `servername` is never used as a
+	// prefix. As this function is concerned with messages send by a
+	// client. In accordance with the ABNF from above, the maximum
+	// length is therefore calculated as:
+	//
+	//   ":" <nick> "!" <host> " "
+	//
+	return 1 + nicklen + 1 + hostlen + 1
+}
+
+func splitPRIVMSG(event *Event, maxLen int) []*Event {
+	newMsg := func(text string) *Event {
+		e := event.Copy()
+		e.Params[len(event.Params)-1] = text
+		return e
+	}
+
+	// Calculate the maximum text message length (excluding targets).
+	rawEvent := event.Copy()
+	rawEvent.Params = event.Params[0:len(event.Params)-1]
+	maxTextLen := maxLen - rawEvent.Len()
+	if (maxTextLen <= 0) {
+		return []*Event{event} // TODO: too long without any text
+	}
+	eventText := event.Last()
+
+	var events []*Event
+	for len(eventText) > maxTextLen {
+		idx := strings.LastIndex(eventText[0:maxTextLen], " ")
+		if idx == -1 {
+			idx = maxTextLen
+		}
+
+		events = append(events, newMsg(eventText[0:idx]))
+		eventText = eventText[idx:]
+	}
+	events = append(events, newMsg(eventText))
+
+	return events
+}
+
+// splitEvent splits a given event into multiple events to satisfy the
+// maximum message length requirements imposed upon the given client by
+// the associated IRC server.
+func splitEvent(client *Client, event *Event) []*Event {
+	// From RFC 2812:
+	//   IRC messages are always lines of characters terminated with a CR-LF
+	//   (Carriage Return - Line Feed) pair, and these messages SHALL NOT
+	//   exceed 512 characters in length, counting all characters including
+	//   the trailing CR-LF.
+	const maxIRClen int = 512 - len("\r\n")
+
+	// We cannot calculate the length of the source part actually
+	// used by the server. Assume the largest possible value and
+	// make sure source is unset for the event to ensure it is not
+	// take into account by `event.Len()`.
+	event.Source = nil // XXX: Operate on a copy instead?
```

Comment:
Should this operate on a deep copy instead? I was under the assumption that messages send by girc itself always satisfy `event.Source == nil` anyhow.